README.md
# Weather Forecast Web Application

This is a weather forecast web application built with Next.js and TypeScript.


## Developing and Running on local Machine

### Requirements

- Node 18+
- Yarn or npm

### Running

First, clone this repository to local file system.

Then in base folder and run the development server

```bash
npm run dev
# or
yarn dev
```

Open [http://localhost:3000](http://localhost:3000) with your browser to preview the application.

### Environment

Provide your own weather bit API `NEXT_PUBLIC_WEATHER_BIT_API_KEY` to overwrite existing sample API key.

Note: This step is optional as sample API key is provided in `.env` file which will valid till 2023-12-03


## Further Considerations for development

| Prioritization | Task Description                                                  |
|----------------|-------------------------------------------------------------------|
| 1              | Create Sever API to call 3rd party API to avoid expose API KEY    |
| 2              | Create docker file to enable running application through docker image |
| 3              | Introduce 3rd party state management tool such as redux           |
| 4              | Setup unit test and mock server                                   |

CHANGELOG.md

# Change Log
All notable changes to this project will be documented in this file.

## [0.2.0] - 2022-11-16

### Added
- Create basic working application for forecasting
### Changed

### Fixed

src/lib/weatherApi/fetchWeatherForecast.ts
import { CurrentWeatherData, DayWeatherForecastData } from "@/interfaces/weather";
import { Location } from "@/interfaces/location";
import { SUCCESS_RESPONSE } from './mockWetherBitForecastApiResponse';
import { Unit } from "@/interfaces/unit";

//TODO: move to env
const USE_MOCK_DATA = false;
const API_TOKEN = process.env.WEATHER_BIT_API_KEY;

const fetchDaily = async (query: string, unit: Unit) => {
  const unitParam = unit === 'metric' ? 'M' : 'I';
  const res = await fetch(`https://api.weatherbit.io/v2.0/forecast/daily?key=${API_TOKEN}&${queryToApiQuery(query)}&days=6&units=${unitParam}`);

  return {
    ok: res.status === 200,
    data: res,
  }
};

const queryToApiQuery = (query: string) => {
  const queryParts = query.split(',');
  if (queryParts.length === 2 && !isNaN(parseFloat(queryParts[0])) && !isNaN(parseFloat(queryParts[1]))) {
    return `lat=${queryParts[0]}&lon=${queryParts[1]}`;
  }

  return `city=${query}`;
}

const responseToLocation = (resp: any): Location => {
  return {
    cityName: resp.city_name,
    countryCode: resp.country_code,
    stateCode: resp.state_code,
  }
}
const responseToWeatherData = (resp: any): CurrentWeatherData => {
  const c = resp.data[0];
    return {
        temperature: c.temp,
        humidity: c.rh,
        windSpeed: c.wind_spd,
        windDirection: c.wind_cdir,
        weatherDescription: c.weather.description,
        weatherIconUrl: `https://cdn.weatherbit.io/static/img/icons/${c.weather.icon}.png`, //TODO: move to icon to to SVG or CSS
    }
}

const responseToForecastData = (resp: any): DayWeatherForecastData[] => {
  const data = resp.data;
  return data.slice(1, 6).map((d: any) => {
    return {
      date: new Date(d.ts * 1000),
      minTemperature: d.min_temp,
      maxTemperature: d.max_temp,
      weatherDescription: d.weather.description,
    }
  })
}

export interface WeatherAndForecastSuccessResponse {
  success: boolean;
  currentWeatherData: CurrentWeatherData;
  location: Location;
  forecasts: DayWeatherForecastData[];
}

export interface WeatherAndForecastErrorResponse {
  success: boolean;
  err_msg: string;
}

export const fetchWeatherAndForecast = async (query: string, unit: Unit): Promise<WeatherAndForecastSuccessResponse|WeatherAndForecastErrorResponse> => {
  if (!query) {
    return {
      success: false,
      err_msg: 'Please provide a location or coordinate to get Started. Please separate latitude and longitude with a comma.'
    };
  }

  let data;
  if (USE_MOCK_DATA) {
    data = SUCCESS_RESPONSE;
  } else {
    const apiResp = await fetchDaily(query, unit);
    if (!apiResp.ok) {
      return {
        success: false,
        err_msg: 'Error fetching weather data, please provide a valid location. If you are using coordinates, please separate latitude and longitude with a comma.'
      };
    } else {
      data = await apiResp.data.json();
    }
  }

  return {
    success: true,
    currentWeatherData: responseToWeatherData(data),
    location: responseToLocation(data),
    forecasts: responseToForecastData(data),
  }
};



src/lib/weatherApi/mockWetherBitForecastApiResponse.ts
export const SUCCESS_RESPONSE = {
  "city_name": "Raleigh",
  "country_code": "US",
  "data": [
    {
      "app_max_temp": 16.5,
      "app_min_temp": 7.7,
      "clouds": 15,
      "clouds_hi": 31,
      "clouds_low": 0,
      "clouds_mid": 0,
      "datetime": "2023-11-13",
      "dewpt": 4.7,
      "high_temp": 15,
      "low_temp": 4.3,
      "max_dhi": null,
      "max_temp": 16.9,
      "min_temp": 7.4,
      "moon_phase": 0.028945,
      "moon_phase_lunation": 0.02,
      "moonrise_ts": 1699877090,
      "moonset_ts": 1699915973,
      "ozone": null,
      "pop": 0,
      "precip": 0,
      "pres": 1014.7,
      "rh": 63,
      "slp": 1025.3,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1699876079,
      "sunset_ts": 1699913248,
      "temp": 11.9,
      "ts": 1699880460,
      "uv": 0,
      "valid_date": "2023-11-13",
      "vis": 14.44,
      "weather": {
        "description": "Few clouds",
        "code": 801,
        "icon": "c02d"
      },
      "wind_cdir": "SW",
      "wind_cdir_full": "southwest",
      "wind_dir": 220,
      "wind_gust_spd": 3.1,
      "wind_spd": 2.1
    },
    {
      "app_max_temp": 17.2,
      "app_min_temp": 2.9,
      "clouds": 42,
      "clouds_hi": 86,
      "clouds_low": 0,
      "clouds_mid": 1,
      "datetime": "2023-11-14",
      "dewpt": 1.3,
      "high_temp": 17.8,
      "low_temp": 3.2,
      "max_dhi": null,
      "max_temp": 17.8,
      "min_temp": 4.3,
      "moon_phase": 0.078553,
      "moon_phase_lunation": 0.05,
      "moonrise_ts": 1699967569,
      "moonset_ts": 1700005394,
      "ozone": 294.4,
      "pop": 0,
      "precip": 0,
      "pres": 1016.8,
      "rh": 56,
      "slp": 1027.5,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1699962539,
      "sunset_ts": 1699999608,
      "temp": 10.4,
      "ts": 1699938060,
      "uv": 3.7,
      "valid_date": "2023-11-14",
      "vis": 12.166,
      "weather": {
        "description": "Broken clouds",
        "code": 803,
        "icon": "c03d"
      },
      "wind_cdir": "SSE",
      "wind_cdir_full": "south-southeast",
      "wind_dir": 168,
      "wind_gust_spd": 4.5,
      "wind_spd": 3
    },
    {
      "app_max_temp": 14.3,
      "app_min_temp": 2.1,
      "clouds": 78,
      "clouds_hi": 100,
      "clouds_low": 0,
      "clouds_mid": 63,
      "datetime": "2023-11-15",
      "dewpt": 3,
      "high_temp": 14.9,
      "low_temp": 5,
      "max_dhi": null,
      "max_temp": 14.9,
      "min_temp": 3.2,
      "moon_phase": 0.151487,
      "moon_phase_lunation": 0.09,
      "moonrise_ts": 1700058042,
      "moonset_ts": 1700095391,
      "ozone": 285,
      "pop": 0,
      "precip": 0,
      "pres": 1019,
      "rh": 68,
      "slp": 1029.8,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1700049000,
      "sunset_ts": 1700085969,
      "temp": 8.9,
      "ts": 1700024460,
      "uv": 3.2,
      "valid_date": "2023-11-15",
      "vis": 17.674,
      "weather": {
        "description": "Overcast clouds",
        "code": 804,
        "icon": "c04d"
      },
      "wind_cdir": "NE",
      "wind_cdir_full": "northeast",
      "wind_dir": 51,
      "wind_gust_spd": 3.1,
      "wind_spd": 2.1
    },
    {
      "app_max_temp": 19,
      "app_min_temp": 4.7,
      "clouds": 58,
      "clouds_hi": 93,
      "clouds_low": 3,
      "clouds_mid": 36,
      "datetime": "2023-11-16",
      "dewpt": 8.6,
      "high_temp": 19.9,
      "low_temp": 10.2,
      "max_dhi": null,
      "max_temp": 19.9,
      "min_temp": 5,
      "moon_phase": 0.151487,
      "moon_phase_lunation": 0.12,
      "moonrise_ts": 1700148283,
      "moonset_ts": 1700181791,
      "ozone": 280.5,
      "pop": 0,
      "precip": 0,
      "pres": 1010.4,
      "rh": 80,
      "slp": 1020.9,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1700135460,
      "sunset_ts": 1700172332,
      "temp": 12.3,
      "ts": 1700110860,
      "uv": 2.6,
      "valid_date": "2023-11-16",
      "vis": 24.128,
      "weather": {
        "description": "Broken clouds",
        "code": 803,
        "icon": "c03d"
      },
      "wind_cdir": "ENE",
      "wind_cdir_full": "east-northeast",
      "wind_dir": 65,
      "wind_gust_spd": 4,
      "wind_spd": 2.7
    },
    {
      "app_max_temp": 20.4,
      "app_min_temp": 10.9,
      "clouds": 72,
      "clouds_hi": 35,
      "clouds_low": 31,
      "clouds_mid": 27,
      "datetime": "2023-11-17",
      "dewpt": 13.4,
      "high_temp": 21.1,
      "low_temp": 11.5,
      "max_dhi": null,
      "max_temp": 21.1,
      "min_temp": 10.2,
      "moon_phase": 0.244715,
      "moon_phase_lunation": 0.16,
      "moonrise_ts": 1700238085,
      "moonset_ts": 1700272254,
      "ozone": 288,
      "pop": 60,
      "precip": 4.76,
      "pres": 999.4,
      "rh": 92,
      "slp": 1009.6,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1700221921,
      "sunset_ts": 1700258697,
      "temp": 14.8,
      "ts": 1700197260,
      "uv": 1.9,
      "valid_date": "2023-11-17",
      "vis": 24.128,
      "weather": {
        "description": "Light rain",
        "code": 500,
        "icon": "r01d"
      },
      "wind_cdir": "ENE",
      "wind_cdir_full": "east-northeast",
      "wind_dir": 78,
      "wind_gust_spd": 6.1,
      "wind_spd": 4
    },
    {
      "app_max_temp": 17.9,
      "app_min_temp": 7.7,
      "clouds": 34,
      "clouds_hi": 44,
      "clouds_low": 70,
      "clouds_mid": 62,
      "datetime": "2023-11-18",
      "dewpt": 8.7,
      "high_temp": 19,
      "low_temp": 5,
      "max_dhi": null,
      "max_temp": 19,
      "min_temp": 6.8,
      "moon_phase": 0.353363,
      "moon_phase_lunation": 0.19,
      "moonrise_ts": 1700327362,
      "moonset_ts": 1700362964,
      "ozone": 300.1,
      "pop": 75,
      "precip": 5.6,
      "pres": 990.8,
      "rh": 76,
      "slp": 1001.1,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1700308381,
      "sunset_ts": 1700345064,
      "temp": 13.2,
      "ts": 1700283660,
      "uv": 3,
      "valid_date": "2023-11-18",
      "vis": 23.04,
      "weather": {
        "description": "Light rain",
        "code": 500,
        "icon": "r01d"
      },
      "wind_cdir": "NW",
      "wind_cdir_full": "northwest",
      "wind_dir": 322,
      "wind_gust_spd": 7.6,
      "wind_spd": 5.1
    },
    {
      "app_max_temp": 15.8,
      "app_min_temp": 2.9,
      "clouds": 1,
      "clouds_hi": 0,
      "clouds_low": 0,
      "clouds_mid": 0,
      "datetime": "2023-11-19",
      "dewpt": -2.9,
      "high_temp": 15.8,
      "low_temp": 2.7,
      "max_dhi": null,
      "max_temp": 15.8,
      "min_temp": 5,
      "moon_phase": 0.47114,
      "moon_phase_lunation": 0.22,
      "moonrise_ts": 1700416155,
      "moonset_ts": 1700453714,
      "ozone": 297.5,
      "pop": 20,
      "precip": 0.04,
      "pres": 996.7,
      "rh": 44,
      "slp": 1007.2,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1700394841,
      "sunset_ts": 1700431432,
      "temp": 9.5,
      "ts": 1700370060,
      "uv": 3.9,
      "valid_date": "2023-11-19",
      "vis": 24.128,
      "weather": {
        "description": "Clear Sky",
        "code": 800,
        "icon": "c01d"
      },
      "wind_cdir": "NW",
      "wind_cdir_full": "northwest",
      "wind_dir": 304,
      "wind_gust_spd": 6.9,
      "wind_spd": 3.6
    },
    {
      "app_max_temp": 12,
      "app_min_temp": 0,
      "clouds": 14,
      "clouds_hi": 18,
      "clouds_low": 0,
      "clouds_mid": 13,
      "datetime": "2023-11-20",
      "dewpt": -6.2,
      "high_temp": 12,
      "low_temp": 4.8,
      "max_dhi": null,
      "max_temp": 12,
      "min_temp": 2.7,
      "moon_phase": 0.590962,
      "moon_phase_lunation": 0.26,
      "moonrise_ts": 1700504578,
      "moonset_ts": 1700457988,
      "ozone": 283.3,
      "pop": 0,
      "precip": 0,
      "pres": 1007.2,
      "rh": 39,
      "slp": 1017.9,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1700481300,
      "sunset_ts": 1700517803,
      "temp": 7.4,
      "ts": 1700456460,
      "uv": 3.9,
      "valid_date": "2023-11-20",
      "vis": 24.128,
      "weather": {
        "description": "Few clouds",
        "code": 801,
        "icon": "c02d"
      },
      "wind_cdir": "W",
      "wind_cdir_full": "west",
      "wind_dir": 265,
      "wind_gust_spd": 5.3,
      "wind_spd": 2.8
    },
    {
      "app_max_temp": 10.1,
      "app_min_temp": 2.9,
      "clouds": 89,
      "clouds_hi": 86,
      "clouds_low": 70,
      "clouds_mid": 65,
      "datetime": "2023-11-21",
      "dewpt": -4.9,
      "high_temp": 10.1,
      "low_temp": 7.2,
      "max_dhi": null,
      "max_temp": 10.1,
      "min_temp": 4.8,
      "moon_phase": 0.705564,
      "moon_phase_lunation": 0.29,
      "moonrise_ts": 1700592757,
      "moonset_ts": 1700548566,
      "ozone": 266.9,
      "pop": 15,
      "precip": 0.3125,
      "pres": 1008.3,
      "rh": 42,
      "slp": 1019,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1700567760,
      "sunset_ts": 1700604175,
      "temp": 7.3,
      "ts": 1700542860,
      "uv": 3.8,
      "valid_date": "2023-11-21",
      "vis": 23.314,
      "weather": {
        "description": "Overcast clouds",
        "code": 804,
        "icon": "c04d"
      },
      "wind_cdir": "ESE",
      "wind_cdir_full": "east-southeast",
      "wind_dir": 111,
      "wind_gust_spd": 4,
      "wind_spd": 2.2
    },
    {
      "app_max_temp": 9.1,
      "app_min_temp": -0.8,
      "clouds": 66,
      "clouds_hi": 86,
      "clouds_low": 60,
      "clouds_mid": 35,
      "datetime": "2023-11-22",
      "dewpt": 1,
      "high_temp": 9.1,
      "low_temp": -1.3,
      "max_dhi": null,
      "max_temp": 9.1,
      "min_temp": 3,
      "moon_phase": 0.808028,
      "moon_phase_lunation": 0.32,
      "moonrise_ts": 1700680809,
      "moonset_ts": 1700639090,
      "ozone": 283.7,
      "pop": 90,
      "precip": 16.8125,
      "pres": 993.8,
      "rh": 70,
      "slp": 1004.4,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1700654219,
      "sunset_ts": 1700690549,
      "temp": 7.1,
      "ts": 1700629260,
      "uv": 3.8,
      "valid_date": "2023-11-22",
      "vis": 17.975,
      "weather": {
        "description": "Heavy rain",
        "code": 502,
        "icon": "r03d"
      },
      "wind_cdir": "WSW",
      "wind_cdir_full": "west-southwest",
      "wind_dir": 243,
      "wind_gust_spd": 7.7,
      "wind_spd": 3.7
    },
    {
      "app_max_temp": 7.4,
      "app_min_temp": -6.1,
      "clouds": 5,
      "clouds_hi": 20,
      "clouds_low": 0,
      "clouds_mid": 0,
      "datetime": "2023-11-23",
      "dewpt": -9.5,
      "high_temp": 7.4,
      "low_temp": 1,
      "max_dhi": null,
      "max_temp": 7.4,
      "min_temp": -1.3,
      "moon_phase": 0.892319,
      "moon_phase_lunation": 0.36,
      "moonrise_ts": 1700768836,
      "moonset_ts": 1700729622,
      "ozone": 298.9,
      "pop": 0,
      "precip": 0,
      "pres": 1000.1,
      "rh": 42,
      "slp": 1010.9,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1700740678,
      "sunset_ts": 1700776925,
      "temp": 2.2,
      "ts": 1700715660,
      "uv": 3.8,
      "valid_date": "2023-11-23",
      "vis": 24.128,
      "weather": {
        "description": "Few clouds",
        "code": 801,
        "icon": "c02d"
      },
      "wind_cdir": "W",
      "wind_cdir_full": "west",
      "wind_dir": 280,
      "wind_gust_spd": 10.2,
      "wind_spd": 5.1
    },
    {
      "app_max_temp": 8.9,
      "app_min_temp": -2.6,
      "clouds": 1,
      "clouds_hi": 3,
      "clouds_low": 0,
      "clouds_mid": 0,
      "datetime": "2023-11-24",
      "dewpt": -6.3,
      "high_temp": 9.8,
      "low_temp": 2,
      "max_dhi": null,
      "max_temp": 9.8,
      "min_temp": 1,
      "moon_phase": 0.953788,
      "moon_phase_lunation": 0.39,
      "moonrise_ts": 1700856938,
      "moonset_ts": 1700820205,
      "ozone": 296.6,
      "pop": 0,
      "precip": 0,
      "pres": 1010,
      "rh": 47,
      "slp": 1020.8,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1700827137,
      "sunset_ts": 1700863303,
      "temp": 5,
      "ts": 1700802060,
      "uv": 3.8,
      "valid_date": "2023-11-24",
      "vis": 24.128,
      "weather": {
        "description": "Clear Sky",
        "code": 800,
        "icon": "c01d"
      },
      "wind_cdir": "NW",
      "wind_cdir_full": "northwest",
      "wind_dir": 318,
      "wind_gust_spd": 8.3,
      "wind_spd": 4
    },
    {
      "app_max_temp": 10.1,
      "app_min_temp": -0.2,
      "clouds": 21,
      "clouds_hi": 81,
      "clouds_low": 3,
      "clouds_mid": 3,
      "datetime": "2023-11-25",
      "dewpt": -5.4,
      "high_temp": 11.1,
      "low_temp": 4.6,
      "max_dhi": null,
      "max_temp": 11.1,
      "min_temp": 2,
      "moon_phase": 0.989577,
      "moon_phase_lunation": 0.43,
      "moonrise_ts": 1700945215,
      "moonset_ts": 1700910837,
      "ozone": 285.9,
      "pop": 0,
      "precip": 0,
      "pres": 1009.3,
      "rh": 45,
      "slp": 1020,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1700913595,
      "sunset_ts": 1700949683,
      "temp": 6.1,
      "ts": 1700888460,
      "uv": 3.7,
      "valid_date": "2023-11-25",
      "vis": 24.128,
      "weather": {
        "description": "Scattered clouds",
        "code": 802,
        "icon": "c02d"
      },
      "wind_cdir": "WNW",
      "wind_cdir_full": "west-northwest",
      "wind_dir": 291,
      "wind_gust_spd": 4.6,
      "wind_spd": 2.5
    },
    {
      "app_max_temp": 11.1,
      "app_min_temp": 3.6,
      "clouds": 9,
      "clouds_hi": 1,
      "clouds_low": 9,
      "clouds_mid": 1,
      "datetime": "2023-11-26",
      "dewpt": -3.7,
      "high_temp": 12.2,
      "low_temp": 6.5,
      "max_dhi": null,
      "max_temp": 12.2,
      "min_temp": 4.6,
      "moon_phase": 0.998846,
      "moon_phase_lunation": 0.46,
      "moonrise_ts": 1701033770,
      "moonset_ts": 1701001436,
      "ozone": 283.4,
      "pop": 0,
      "precip": 0,
      "pres": 1007.5,
      "rh": 44,
      "slp": 1018.2,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1701000053,
      "sunset_ts": 1701036065,
      "temp": 8.1,
      "ts": 1700974860,
      "uv": 3.7,
      "valid_date": "2023-11-26",
      "vis": 24.128,
      "weather": {
        "description": "Few clouds",
        "code": 801,
        "icon": "c02d"
      },
      "wind_cdir": "NNW",
      "wind_cdir_full": "north-northwest",
      "wind_dir": 344,
      "wind_gust_spd": 3.2,
      "wind_spd": 1.6
    },
    {
      "app_max_temp": 12.2,
      "app_min_temp": 6.1,
      "clouds": 100,
      "clouds_hi": 100,
      "clouds_low": 57,
      "clouds_mid": 100,
      "datetime": "2023-11-27",
      "dewpt": 2.6,
      "high_temp": 13.4,
      "low_temp": 3.6,
      "max_dhi": null,
      "max_temp": 13.4,
      "min_temp": 6.5,
      "moon_phase": 0.982749,
      "moon_phase_lunation": 0.49,
      "moonrise_ts": 1701122697,
      "moonset_ts": 1701091840,
      "ozone": 291.3,
      "pop": 30,
      "precip": 0.8125,
      "pres": 1003.5,
      "rh": 62,
      "slp": 1014.1,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1701086510,
      "sunset_ts": 1701122448,
      "temp": 9.7,
      "ts": 1701061260,
      "uv": 3.7,
      "valid_date": "2023-11-27",
      "vis": 24.128,
      "weather": {
        "description": "Overcast clouds",
        "code": 804,
        "icon": "c04d"
      },
      "wind_cdir": "WSW",
      "wind_cdir_full": "west-southwest",
      "wind_dir": 250,
      "wind_gust_spd": 4.8,
      "wind_spd": 2.8
    },
    {
      "app_max_temp": 12.4,
      "app_min_temp": 2.2,
      "clouds": 38,
      "clouds_hi": 2,
      "clouds_low": 0,
      "clouds_mid": 38,
      "datetime": "2023-11-28",
      "dewpt": -1.9,
      "high_temp": 13.6,
      "low_temp": 7.8,
      "max_dhi": null,
      "max_temp": 13.6,
      "min_temp": 3.6,
      "moon_phase": 0.94409,
      "moon_phase_lunation": 0.53,
      "moonrise_ts": 1701212046,
      "moonset_ts": 1701181861,
      "ozone": 313.3,
      "pop": 0,
      "precip": 0,
      "pres": 1004.8,
      "rh": 52,
      "slp": 1015.4,
      "snow": 0,
      "snow_depth": 0,
      "sunrise_ts": 1701172967,
      "sunset_ts": 1701208834,
      "temp": 8.2,
      "ts": 1701147660,
      "uv": 3.6,
      "valid_date": "2023-11-28",
      "vis": 24.128,
      "weather": {
        "description": "Scattered clouds",
        "code": 802,
        "icon": "c02d"
      },
      "wind_cdir": "ESE",
      "wind_cdir_full": "east-southeast",
      "wind_dir": 107,
      "wind_gust_spd": 3.8,
      "wind_spd": 2.2
    }
  ],
  "lat": "35.7721",
  "lon": "-78.63861",
  "state_code": "NC",
  "timezone": "America/New_York"
};